src/librustc_mir/build/block.rs

```diff
@@ -143,7 +143,13 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
         if let Some(expr) = expr {
             unpack!(block = this.into(destination, block, expr));
         } else {
-            this.cfg.push_assign_unit(block, source_info, destination);
+            let tcx = this.hir.tcx();
+            let ty = destination.ty(&this.local_decls, tcx).to_ty(tcx);
+            if ty.is_nil() {
```

Comment:
Just to be sure I understand, the premise here is that the type is either `()` or else the block must diverge, since any other type would have yielded a compilation error?

Reply:
Yeah, with the current semantics, any other type should be impossible (i.e. any other type must be explicit and go through the `Some(expr)` branch instead) , because the only implicit return types are `()` and `!`.

Reply:
Could we have a comment? That's a bit non-obvious. (If the code were like `if ty.is_never()` it'd be more obvious...)

Reply:
Sure. (The reason I opted for `ty.is_nil()` rather than `ty.is_never()` here is because it seemed less contextual — it's always true that we only want to assign a unit if the type is unit. One alternative to make it absolutely explicit would be to check for `!ty.is_never()` in an `else` clause and throw a warning there that it should be impossible, but this seems like overkill.)